lib/spack/spack/url_buildcache.py

```diff
@@ -1335,25 +1336,35 @@ class MirrorURLAndVersion:
     url: str
     version: int
 
-    def __init__(self, url: str, version: int):
+    def __init__(self, url: str, version: int, view: Optional[str] = None):
         self.url = url
         self.version = version
+        self.view = view
 
     def __str__(self):
-        return f"{self.url}__v{self.version}"
+        s = f"{self.url}__v{self.version}"
+        if self.view:
+            s += f"__@{self.view}"
+        return s
 
     def __eq__(self, other):
         if isinstance(other, MirrorURLAndVersion):
-            return self.url == other.url and self.version == other.version
+            return all(
+                [self.url == other.url, self.version == other.version, self.view == other.view]
+            )
         return False
 
     def __hash__(self):
-        return hash((self.url, self.version))
+        return hash((self.url, self.version, self.view))
 
     @classmethod
     def from_string(cls, s: str):
-        parts = s.split("__v")
-        return cls(parts[0], int(parts[1]))
+        m = re.match(r"^(.*)__v([0-9]+)(__@(.*))?$", s)
```

Comment:
Use non-capturing groups for things you don't need

```suggestion
        m = re.match(r"^(.*)__v([0-9]+)(?:__@(.*))?$", s)
```

Reply:
Neat, didn't know that existed!